designs/0035-cmdstanrs.md

```diff
@@ -0,0 +1,348 @@
+Emacs 28.2, one can view options at each node in the argument tree by
+code that looks something like the following:
+
+```rust
+ArgumentTree::builder(). // hover on the `.`
+```
+
+If one has a Rust language server and completion support in their
+editor, this is a free side effect. Whether it will help anyone
+is uncertain.
+
+### Coverage
+
+The objective is for the interface to cover all options which can be
+passed to a compiled Stan program, that is, all methods and all
+options for said methods.
+
+## Separation of concerns
+
+Other than the argument tree support, the interface proposed is very
+simple: the user can compile a Stan program, call it with arguments,
+get basic information from the output, and call
+`diagnose`/`stansummary`.
+
+Below, I provide the rationale for exclusion of two aspects. My
+judgment is that they are useful, but are best developed separately.
+
+### Serialization
+
+It is trivial to provide a function such as 
+```rust
+fn write_json<T: Serialize>(data: &T, file: &Path) {}
+```
+but it serves no purpose -- it does not enforce the conventions
+adopted for representing Stan data types in JSON (e.g. matrices
+represented as a vector of *row* vectors, not a vector of *column*
+vectors), hence, would likely lead to unexpected (and potentially
+silent!) errors.
+
+In order to develop a serializer which respects the conventions for
+Stan data types, one would need to declare conventions for the mapping
+of Rust data types to Stan data
+types. [serde_json](https://github.com/serde-rs/json) would be nice to
+use, but has some incompatibilities (Rust tuple is represented as an
+array, rather than an object).
+
+Moreover, to represent matrices, complex numbers, etc., one would need
+to support types from the crate ecosystem since the standard library
+lacks these -- [nalgebra](https://github.com/dimforge/nalgebra) and
+[num-complex](https://github.com/rust-num/num-complex) are reasonable
+choices, but nonetheless represent decisions to be made!
+
+From a design perspective, this is a great place to defer to the user,
+at least for the moment. A principled approach would involve writing a
+data format for [serde](https://serde.rs/).
+
+### Deserialization
+
+Parsing Stan CSVs to a strongly-typed representation is simple if one
+wishes to simply obtain a matrix of values (or `Vec<Vec<f64>>` if we
+limit ourselves to `std` library types). However, one needs to extract
+the variables from each line, thus, one needs to know the types (and
+their dimensions). A recursive definition of types using `enum`s and
+`struct`s could probably work to represent such a thing in Rust, but
+may not necessarily be particularly ergonomic (i.e. much unavoidable
+boilerplate would be needed to use the resultant type).
+
+Procedural macros, applied to a Stan program stored in a string
+literal in a Rust program, could be used to generate types and a
+parser for Stan CSVs produced said program. However, in order to
```

Comment:
There is a middle-ground between this (very cool, but tricky) idea and doing nothing, which is providing some primitives (which may themselves be macros) that allow the user to provide the information required about types and names of variables. Worth considering, and it seems like any future automation could build on it.

Reply:
The gate here is: decide on conventions for representation of Stan types as Rust types. `nalgebra` is quite similar to Eigen, including memory layout, so matrices are covered.

For arrays of dimension greater than 2, there is a problem.
- Yes, there is `ndarray`, but its [`future is uncertain`](https://github.com/rust-ndarray/ndarray/issues/1272). 
- Nested `Vec<Vec<Vec ... >>>` naturally works, but I doubt users would like it (no syntactic sugar for indexing).

If we take the latter option (the only sensible course for stable development), then it would be relatively straightforward to use a procedural macro to derive a parser for a user-provided struct where the field names and respective types must match the Stan CSV output. I do something similar for my use cases.